internal/cutil/iovec.go

```diff
@@ -2,77 +2,60 @@ package cutil
 
 /*
 #include <stdlib.h>
+#include <string.h>
 #include <sys/uio.h>
 */
 import "C"
-
 import (
 	"unsafe"
 )
 
-var iovecSize uintptr
-
-// StructIovecPtr is an unsafe pointer wrapping C's `*struct iovec`.
-type StructIovecPtr unsafe.Pointer
-
-// Iovec helps manage struct iovec arrays needed by some C functions.
+// Iovec is a slice of iovec structs. Might have allocated C memory, so it must
+// be freed with the Free() method.
 type Iovec struct {
-	// cvec represents an array of struct iovec C memory
-	cvec unsafe.Pointer
-	// length of the array (in elements)
-	length int
+	iovec []C.struct_iovec
+	data  [][]byte
 }
 
-// NewIovec creates an Iovec, and underlying C memory, of the specified size.
-func NewIovec(l int) *Iovec {
-	r := &Iovec{
-		cvec:   C.malloc(C.size_t(l) * C.size_t(iovecSize)),
-		length: l,
+// ByteSlicesToIovec creates an Iovec and links it to Go buffers in data.
+func ByteSlicesToIovec(data [][]byte) (v Iovec) {
+	v.iovec = make([]C.struct_iovec, len(data))
+	v.data = data
+	for i, b := range v.data {
+		v.iovec[i].iov_base = C.CBytes(b)
+		v.iovec[i].iov_len = C.size_t(len(b))
 	}
-	return r
+	return
 }
 
-// ByteSlicesToIovec takes a slice of byte slices and returns a new iovec that
-// maps the slice data to struct iovec entries.
-func ByteSlicesToIovec(data [][]byte) *Iovec {
-	iov := NewIovec(len(data))
-	for i := range data {
-		iov.Set(i, data[i])
+// SyncToData writes the iovec buffers into the linked Go buffers
+func (v Iovec) SyncToData() {
```

Comment:
Again, this feels a bit like its exposing too much detail about the implementation to the external users. What about an api like: 
```go
// Data synchronizes the internal structures in the Iovec, returning an up-to-date representation of the data as a slice of byte-slices.
func (v Iovec) Data() [][]byte {
   // ... stuff goes here....
}
```

Reply:
Err. now that I've posted my suggestion, now I wonder if a better name is the verb rather than the noun. IOW, `Sync` rather than `Data`.  :-)

Reply:
I'll incorporate your comments in #441 if you don't mind (they are still valid there).